File: easy/0104_Max_Depth_BT_Iterative.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        stack = [(root, 1)]

        max_depth = 0

        while stack:
            node, depth = stack.pop()
            if node.left:
                stack.append((node.left, depth+1))
            if node.right:
                stack.append((node.right, depth+1))
            max_depth = max(max_depth, depth)
        
        return max_depth

File: easy/0104_Max_Depth_BT_Iterative.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int maxDepth(struct TreeNode* root) {
    if(root == NULL){
        return 0;
    }
    int depth[10000];
    struct TreeNode *stack[10000];
    int max_depth = 0, top = 0;
    
    stack[top] = root; 
    depth[top] = 1; 
    top++; 
    while(top > 0){
        top--; 
        struct TreeNode *node = stack[top];
        int curr_depth = depth[top];

        if(node->left != NULL){
            stack[top] = node->left;
            depth[top] = curr_depth +1;
            top++; 
            
        }
        if(node->right != NULL){
            stack[top] = node->right;
            depth[top] = curr_depth +1;
            top++;
        }
        if(max_depth< curr_depth){
            max_depth = curr_depth;
        }
    }

    return max_depth;

    }
